action_server/frontend/src/views/actions/components/ActionRun.tsx

```diff
@@ -128,13 +132,45 @@ export const ActionRun: FC<Props> = ({ action, actionPackage }) => {
       runAction({
         actionPackageName: toKebabCase(actionPackage.name),
         actionName: toKebabCase(action.name),
-        args: formDatatoPayload(formData),
+        args: useRawJSON ? JSON.parse(formRawJSON) : formDataToPayload(formData),
         apiKey: serverConfig?.auth_enabled ? apiKey : undefined,
       });
     },
-    [action, actionPackage, formData, apiKey, serverConfig],
+    [action, actionPackage, formData, apiKey, serverConfig, useRawJSON, formRawJSON],
   );
 
+  const onCodeChange = useCallback(
+    (value: string) => {
+      setFormRawJSON(value);
+      try {
+        JSON.parse(value) as Payload;
+        setErrorJSON(undefined);
+      } catch (_) {
+        setErrorJSON('Error while parsing JSON. Please review value and try again');
+      }
+    },
+    [setFormRawJSON, setErrorJSON],
+  );
+
+  useEffect(() => {
+    setFormData(propertiesToFormData(JSON.parse(action.input_schema)));
+  }, [action, actionPackage]);
+
+  useEffect(() => {
+    if (formData.length === 0) {
+      setFormRawJSON('No input data...');
+      return;
+    }
+    const payload = formDataToPayload(formData);
+    setFormRawJSON(JSON.stringify(payload, null, 4));
+  }, [formData]);
+
+  useEffect(() => {
+    if (!useRawJSON && formRawJSON && formData.length >= 0) {
+      setFormData(payloadToFormData(JSON.parse(formRawJSON), formData));
+    }
+  }, [useRawJSON]);
```

Comment:
I think `setFormRawJSON` can be affected by `useRawJSON` not on every formData change,  so both of these effects can be merged:

```suggestion
  useEffect(() => {
    if (!useRawJSON){
      setFormData(payloadToFormData(JSON.parse(formRawJSON), formData));
    } else {
      const payload = formDataToPayload(formData);
      setFormRawJSON(JSON.stringify(payload, null, 4));
    }
  }, [useRawJSON]);
```